lib/surface/compiler/eex_engine.ex

```diff
@@ -216,124 +216,188 @@ defmodule Surface.Compiler.EExEngine do
     end
   end
 
-  defp to_expression(
-         %AST.Component{
-           module: module,
-           type: Surface.LiveView,
-           props: props
-         } = component,
-         _buffer,
-         _state
-       ) do
-    props_expr =
-      collect_component_props(module, props)
-      |> Enum.reject(fn {_, value} -> is_nil(value) end)
+  # Dynamic component
+  defp to_expression(%AST.FunctionComponent{type: :dynamic} = component, buffer, state) do
+    %AST.FunctionComponent{
+      module: %AST.AttributeExpr{value: module_expr},
+      fun: fun,
+      props: props,
+      meta: meta
+    } = component
+
+    fun_expr =
+      case fun do
+        nil -> :render
+        %AST.AttributeExpr{value: expr} -> expr
+      end
+
+    {props_expr, dynamic_props_expr} = build_props_expressions(nil, component)
+    {context_expr, context_var, state} = process_context(nil, nil, props, meta.caller, state)
+    {do_block, slot_meta, slot_props} = collect_slot_meta(component, buffer, state, context_var)
 
     quote generated: true do
-      live_render(
-        @socket,
-        unquote(module),
-        unquote(props_expr)
+      component(
+        &apply(unquote(module_expr), unquote(fun_expr), [&1]),
+        Surface.build_assigns(
+          unquote(context_expr),
+          unquote(props_expr),
+          unquote(dynamic_props_expr),
+          unquote(slot_props),
+          unquote(slot_meta),
+          unquote(module_expr),
+          unquote(meta.node_alias)
+        ),
+        unquote(do_block)
       )
     end
     |> maybe_print_expression(component)
   end
 
-  defp to_expression(
-         %ast_type{
-           module: module,
-           type: component_type,
-           props: props,
-           dynamic_props: dynamic_props,
-           templates: templates,
-           meta: meta
-         } = component,
-         buffer,
-         state
-       )
-       when ast_type in [AST.Component, AST.SlotableComponent] do
-    props_expr = collect_component_props(module, props)
+  # Local function component
+  defp to_expression(%AST.FunctionComponent{type: :local} = component, buffer, state) do
+    %AST.FunctionComponent{module: module, fun: fun, props: props, meta: meta} = component
 
-    dynamic_props_expr = handle_dynamic_props(dynamic_props)
+    {props_expr, dynamic_props_expr} = build_props_expressions(nil, component)
+    {context_expr, context_var, state} = process_context(module, fun, props, meta.caller, state)
+    {do_block, slot_meta, slot_props} = collect_slot_meta(component, buffer, state, context_var)
 
-    caller_component_type = Module.get_attribute(meta.caller.module, :component_type)
+    quote generated: true do
+      component(
+        &(unquote(Macro.var(fun, __MODULE__)) / 1),
```

Comment:
The spacing here looks odd, but I assume that's not an actual issue?

Reply:
Yeah, that's not an issue. I find that ugly too but that's the way the formatter handles it so there's not much we can do about it, I guess :)